src/server/game/Entities/Creature/Trainer.cpp

```diff
@@ -0,0 +1,243 @@
+        {
+            SendTeachFailure(npc, player, spellId, FailReason::NotEnoughSkill);
+            return;
+        }
+
+        float reputationDiscount = player->GetReputationPriceDiscount(npc);
+        int32 moneyCost = int32(trainerSpell->MoneyCost * reputationDiscount);
+        if (!player->HasEnoughMoney(moneyCost))
+        {
+            SendTeachFailure(npc, player, spellId, FailReason::NotEnoughMoney);
+            return;
+        }
+
+        player->ModifyMoney(-moneyCost);
+
+        npc->SendPlaySpellVisual(179);
+        npc->SendPlaySpellImpact(player->GetGUID(), 362);
+
+        // learn explicitly or cast explicitly
+        if (trainerSpell->IsCastable())
+            player->CastSpell(player, trainerSpell->SpellId, true);
+        else
+            player->LearnSpell(trainerSpell->SpellId, false);
+
+        SendTeachSucceeded(npc, player, spellId);
+    }
+
+    Spell const* Trainer::GetSpell(uint32 spellId) const
+    {
+        auto itr = std::find_if(_spells.begin(), _spells.end(), [spellId](Spell const& trainerSpell)
+        {
+            return trainerSpell.SpellId == spellId;
+        });
+
+        if (itr != _spells.end())
+            return &(*itr);
+
+        return nullptr;
+    }
+
+    bool Trainer::CanTeachSpell(Player const* player, Spell const* trainerSpell) const
+    {
+        SpellState state = GetSpellState(player, trainerSpell);
+        if (state != SpellState::Available)
+            return false;
+
+        SpellInfo const* trainerSpellInfo = sSpellMgr->AssertSpellInfo(trainerSpell->SpellId);
+        if (trainerSpellInfo->IsPrimaryProfessionFirstRank() && !player->GetFreePrimaryProfessionPoints())
+            return false;
+
+        return true;
+    }
+
+    SpellState Trainer::GetSpellState(Player const* player, Spell const* trainerSpell) const
+    {
+        if (player->HasSpell(trainerSpell->SpellId))
+            return SpellState::Known;
+
+        // check race/class requirement
+        if (!player->IsSpellFitByClassAndRace(trainerSpell->SpellId))
+            return SpellState::Unavailable;
+
+        // check skill requirement
+        if (trainerSpell->ReqSkillLine && player->GetBaseSkillValue(trainerSpell->ReqSkillLine) < trainerSpell->ReqSkillRank)
+            return SpellState::Unavailable;
+
+        for (int32 reqAbility : trainerSpell->ReqAbility)
+            if (reqAbility && !player->HasSpell(reqAbility))
+                return SpellState::Unavailable;
+
+        // check level requirement
+        if (player->getLevel() < trainerSpell->ReqLevel)
+            return SpellState::Unavailable;
+
+        // check ranks
+        bool hasLearnSpellEffect = false;
+        bool knowsAllLearnedSpells = true;
+        for (uint8 i = 0; i < MAX_SPELL_EFFECTS; ++i)
+        {
+            SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(trainerSpell->SpellId);
+            if (!spellInfo || !spellInfo->Effects[i].Effect == SPELL_EFFECT_LEARN_SPELL)
```

Comment:
Incorrect condition here - `!spellInfo->Effects[i].Effect == SPELL_EFFECT_LEARN_SPELL`, this should be `spellInfo->Effects[i].Effect != SPELL_EFFECT_LEARN_SPELL`